services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java

```diff
@@ -0,0 +1,326 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.protobuf.service;
+
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.Descriptors.FileDescriptor;
+import com.google.protobuf.Descriptors.MethodDescriptor;
+import com.google.protobuf.Descriptors.ServiceDescriptor;
+
+import io.grpc.NotifyOnServerBuild;
+import io.grpc.Server;
+import io.grpc.ServerServiceDefinition;
+import io.grpc.Status;
+import io.grpc.protobuf.ProtoFileDescriptorSupplier;
+import io.grpc.reflection.v1alpha.ErrorResponse;
+import io.grpc.reflection.v1alpha.ExtensionNumberResponse;
+import io.grpc.reflection.v1alpha.ExtensionRequest;
+import io.grpc.reflection.v1alpha.FileDescriptorResponse;
+import io.grpc.reflection.v1alpha.ListServiceResponse;
+import io.grpc.reflection.v1alpha.ServerReflectionGrpc;
+import io.grpc.reflection.v1alpha.ServerReflectionRequest;
+import io.grpc.reflection.v1alpha.ServerReflectionResponse;
+import io.grpc.reflection.v1alpha.ServiceResponse;
+import io.grpc.stub.StreamObserver;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * Provides a reflection service for Protobuf services (including the reflection service itself).
+ */
+public final class ProtoReflectionService extends ServerReflectionGrpc.ServerReflectionImplBase
+    implements NotifyOnServerBuild {
+  private Server server;
+  private Set<String> serviceNames;
+  private Map<String, FileDescriptor> fileDescriptorsByName;
+  private Map<String, FileDescriptor> fileDescriptorsBySymbol;
+  private Map<String, Map<Integer, FileDescriptor>> fileDescriptorsByExtensionAndNumber;
+  private Boolean mapsInitialized = false;
+
+  /**
+   * Retrieves the services registered to the server at build time, and stores those services with a
+   * {@link ProtoFileDescriptorSupplier}.
+   */
+  public void notifyOnBuild(Server server) {
+    this.server = server;
+  }
+
+  private void processExtension(FieldDescriptor extension, FileDescriptor fd) {
+    String extensionName = extension.getContainingType().getFullName();
+    int extensionNumber = extension.getNumber();
+    if (fileDescriptorsByExtensionAndNumber.containsKey(extensionName)) {
```

Comment:
can this if check be reversed?  Basically:

```
if (absent) {
  put empty map
}
insert entry
```


That reduce indentation and avoids copying code.